pkg/rpc/rpc.go

```diff
@@ -0,0 +1,148 @@
+package rpc
+
+import (
+	"encoding/json"
+	"fmt"
+	"github.com/icinga/icinga-notifications/pkg/plugin"
+	"go.uber.org/zap"
+	"io"
+	"sync"
+)
+
+type Error struct {
+	cause error
+}
+
+func (err *Error) Error() string {
+	return fmt.Sprintf("RPCError: %s", err.cause.Error())
+}
+
+func (err *Error) Unwrap() error {
+	return err.cause
+}
+
+type RPC struct {
+	writer  io.Closer // use encoder for writing instead
+	encoder *json.Encoder
+	decoder *json.Decoder
+	logger  *zap.SugaredLogger
+
+	pendingRequests map[uint64]chan plugin.JsonRpcResponse
+	lastRequestId   uint64
+
+	errChannel chan error
+	rpcErr     *Error
+
+	requestsMu sync.Mutex
+	encoderMu  sync.Mutex
+	errMu      sync.Mutex
+}
+
+func NewRPC(writer io.WriteCloser, reader io.Reader, logger *zap.SugaredLogger) *RPC {
+	rpc := &RPC{
+		writer:          writer,
+		encoder:         json.NewEncoder(writer),
+		decoder:         json.NewDecoder(reader),
+		pendingRequests: map[uint64]chan plugin.JsonRpcResponse{},
+		logger:          logger,
+		errChannel:      make(chan error, 1),
+	}
+
+	go rpc.processResponses()
+
+	return rpc
+}
+
+func (r *RPC) Call(method string, params json.RawMessage) (json.RawMessage, error) {
+	if err := r.getErr(); err != nil {
+		return nil, err
+	}
+
+	promise := make(chan plugin.JsonRpcResponse, 1)
+
+	r.requestsMu.Lock()
+	r.lastRequestId++
+	newId := r.lastRequestId
+	r.pendingRequests[newId] = promise
+	r.requestsMu.Unlock()
+
+	r.encoderMu.Lock()
+	err := r.encoder.Encode(plugin.JsonRpcRequest{Method: method, Params: params, Id: newId})
+	r.encoderMu.Unlock()
+	if err != nil {
+		err = fmt.Errorf("failed to write request: %w", err)
+		r.setErr(err)
+
+		return nil, err
+	}
+
+	select {
+	case response := <-promise:
+		if response.Error != "" {
+			return nil, fmt.Errorf("plugin response contains error: %s", response.Error)
+		}
+
+		return response.Result, nil
+
+	case <-r.errChannel:
+		return nil, r.getErr()
+	}
+}
+
+func (r *RPC) processResponses() {
+	defer func() {
+		r.requestsMu.Lock()
+		r.pendingRequests = nil
+		r.requestsMu.Unlock()
+	}()
+
+	for r.getErr() == nil {
+		var response plugin.JsonRpcResponse
+		err := r.decoder.Decode(&response)
+
+		if err != nil {
+			r.logger.Infof("dropping %d pending request(s)", len(r.pendingRequests))
+
+			r.setErr(fmt.Errorf("failed to decode json response: %w", err))
+
+			return
+		}
+
+		r.requestsMu.Lock()
+		promise := r.pendingRequests[response.Id]
+		delete(r.pendingRequests, response.Id)
+		r.requestsMu.Unlock()
+
+		if promise != nil {
+			promise <- response
+		} else {
+			r.logger.Warn("Unknown response")
```

Comment:
Log message could be a bit more informative. Something like "Ignored response for unknown ID $id/without ID".